velox/exec/tests/HashJoinTest.cpp

```diff
@@ -8243,4 +8243,184 @@ TEST_F(HashJoinTest, nanKeys) {
+   public:
+    BlockedOperatorFactory() = default;
+
+    std::unique_ptr<Operator> toOperator(
+        DriverCtx* ctx,
+        int32_t id,
+        const core::PlanNodePtr& node) override {
+      if (std::dynamic_pointer_cast<const BlockedNode>(node)) {
+        return std::make_unique<BlockedOperator>(ctx, id, node, blockedCb_);
+      }
+      return nullptr;
+    }
+
+    void setBlockedCb(BlockedCb blockedCb) {
+      blockedCb_ = std::move(blockedCb);
+    }
+
+   private:
+    BlockedCb blockedCb_{nullptr};
+  };
+
+  auto blockedOperatorFactoryUniquePtr =
+      std::make_unique<BlockedOperatorFactory>();
+  auto blockedOperatorFactory = blockedOperatorFactoryUniquePtr.get();
+  Operator::registerOperator(std::move(blockedOperatorFactoryUniquePtr));
+
+  std::vector<ContinuePromise> unblockPromises;
+  std::vector<std::shared_ptr<bool>> shouldBlocks;
+  blockedOperatorFactory->setBlockedCb(
+      [&](ContinuePromise&& promise, std::shared_ptr<bool> shouldBlock) {
+        unblockPromises.push_back(std::move(promise));
+        shouldBlocks.push_back(shouldBlock);
+      });
+
+  folly::EventCount arbitrationWait;
+  std::atomic<bool> arbitrationWaitFlag{true};
+  ::facebook::velox::common::testutil::ScopedTestValue _scopedTestValue15(
+      "facebook::velox::exec::HashBuild::finishHashBuild",
+      std::function<void(exec::HashBuild*)>([&](Operator* /* unused */) {
+        arbitrationWaitFlag = false;
+        arbitrationWait.notifyAll();
+      }));
+  std::thread arbitrationThread([&]() {
+    arbitrationWait.await([&]() { return !arbitrationWaitFlag.load(); });
+    memory::memoryManager()->shrinkPools();
+    for (auto& shouldBlock : shouldBlocks) {
+      *shouldBlock = false;
+    }
+    for (auto& unblockPromise : unblockPromises) {
+      unblockPromise.setValue();
+    }
+  });
+
+  auto rowType = ROW({{"c0", INTEGER()}, {"c1", INTEGER()}});
+  std::vector<RowVectorPtr> vectors = createVectors(1, rowType, fuzzerOpts_);
+  createDuckDbTable(vectors);
+  auto planNodeIdGenerator = std::make_shared<core::PlanNodeIdGenerator>();
+  const auto spillDirectory = exec::test::TempDirectoryPath::create();
+  auto plan = PlanBuilder(planNodeIdGenerator)
+                  .values(vectors)
+                  .project({"c0 AS t0", "c1 AS t1"})
+                  .hashJoin(
+                      {"t0"},
+                      {"u0"},
+                      PlanBuilder(planNodeIdGenerator)
+                          .values(vectors)
+                          .project({"c0 AS u0", "c1 AS u1"})
+                          .planNode(),
+                      "",
+                      {"t1"},
+                      core::JoinType::kInner)
+                  .addNode([&](std::string id, core::PlanNodePtr input) {
+                    return std::make_shared<BlockedNode>(id, input);
+                  })
+                  .planNode();
+
+  std::shared_ptr<core::QueryCtx> joinQueryCtx =
```

Comment:
Do we need to customized query ctx here?

Reply:
Just a plain query ctx, not customized, to feed to AssertQueryBuilder